File: internal/config/hermes_fingerprint_test.go
// hermes_fingerprint_test.go — system-wide guard tests for Hermes code injection
// points in upstream-shared Go files. Upstream merges can silently drop injected
// code when git merge doesn't flag a conflict (text around the injection changes
// but the injected block itself doesn't overlap — merge drops it cleanly).
//
// Documented losses:
//   - sqz CompressToolOutput wiring in boot.go (lost, fixed h52)
//   - SettingsView Hotbar/Profiles fields in settings_app.go (lost, fixed h52)
//   - render.go Hermes TOML sections (lost, fixed h11)
//
// Every fingerprint was verified against live code before commit.
// If any check fails, an upstream merge just destroyed Hermes code.

package config

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestHermesFingerprintsGo(t *testing.T) {
	repoRoot := findHermesRepoRoot(t)

	// Each entry: file path relative to repo root → required substrings.
	checks := map[string][]string{
		// ── internal/boot/boot.go — Hermes agent wiring ──
		"internal/boot/boot.go": {
			"agentlog.Init",        // operational logging init
			"applyExchangeRate",    // CNY→USD exchange rate
			"CompressToolOutput",   // sqz compressor wiring
			"VisionProv",           // auxiliary vision provider
			"ctrl.SetMesh",         // mesh council wiring
			"ctrl.SetLearner",      // learner pattern detection
		},

		// ── internal/control/controller.go — Hermes struct fields ──
		// Uses raw string literals because controller.go uses tab indentation.
		"internal/control/controller.go": {
			`reasonix/internal/mesh"`,
			`reasonix/internal/scheduler"`,
			`reasonix/internal/learn"`,
			"\tschedule ", // tab-indented field (line 92-ish)
			"\tmesh ",     // tab-indented field (line 93-ish)
			"\tlearner ",  // tab-indented field (line 94-ish)
			"Schedule *scheduler.Scheduler",
			"Mesh *mesh.Mesh",
		},

		// ── internal/config/config.go — Hermes config types ──
		"internal/config/config.go": {
			"HotbarConfig struct",       // hotbar config type
			"CompressToolOutputEnabled", // compress enabled getter
			"ActiveProfile",             // active profile field
			"BillingConfig",             // billing config type
			"LearnConfig",               // learn config type
			"MeshConfig",                // mesh config type
			"ScheduleConfig",            // schedule config type
			"CollabConfig",              // collab config type
			"MarketplaceConfig",         // marketplace config type
			"EmbeddingConfig",           // embedding config type
			"AgentLogConfig",            // agent log config type
			"DiscordBotConfig",          // Discord bot config type
			"TelegramBotConfig",         // Telegram bot config type
			"LineBotConfig",             // LINE bot config type
			"SlackBotConfig",            // Slack bot config type
		},

		// ── internal/config/render.go — Hermes TOML section rendering ──
		"internal/config/render.go": {
			"[desktop.hotbar]",
			"[schedule]",
			"[learn]",
			"[mesh]",
			"[bot.discord]",
			"[bot.telegram]",
			"[bot.line]",
			"[bot.slack]",
			"active_profile",
		},

		// ── internal/config/edit_test.go — Hermes theme test ──
		"internal/config/edit_test.go": {
			`"Hermes", "hermes"`,
		},

		// ── desktop/settings_app.go — Hermes SettingsView + Wails bindings ──
		"desktop/settings_app.go": {
			"// Hermes —",
			"SetDesktopHotbar",
			"SetProfiles",
			"func hotbarView(",
		},

		// ── desktop/main.go — branding ──
		"desktop/main.go": {
			`"Reasonix-Hermes"`,
		},

		// ── desktop/tray.go — gold Hermes tray icon ──
		"desktop/tray.go": {
			"isHermesTheme",
		},
	}

	for file, fingerprints := range checks {
		path := filepath.Join(repoRoot, file)
		data, err := os.ReadFile(path)
		if err != nil {
			t.Errorf("%s: cannot read: %v", file, err)
			continue
		}
		content := string(data)
		for _, fp := range fingerprints {
			if !strings.Contains(content, fp) {
				t.Errorf("%s: MISSING %q — upstream merge may have dropped Hermes code", file, fp)
			}
		}
	}
}

func findHermesRepoRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatalf("os.Getwd: %v", err)
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatal("cannot find repo root (no go.mod found)")
		}
		dir = parent
	}
}

File: desktop/frontend/src/__tests__/hermes-fingerprint.test.ts
// Run: tsx src/__tests__/hermes-fingerprint.test.ts
//
// SYSTEM-WIDE guard: verifies every Hermes code block injected into
// upstream-shared files still exists. Upstream merges can silently
// drop injected code when git merge doesn't flag a conflict.
//
// Documented losses:
//   - hotbar handler in App.tsx (lost 9 days, Jun 12–21 2026, fixed b6de38ac)
//   - sqz CompressToolOutput wiring in boot.go (lost, fixed h52)
//   - SettingsView Hotbar/Profiles fields (lost, fixed h52)
//   - render.go Hermes TOML sections (lost, fixed h11)
//
// Every fingerprint was verified against the live codebase before commit.
// If a fingerprint fails, an upstream merge just destroyed Hermes code.

import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

let passed = 0;
let failed = 0;

function ok(cond: boolean, label: string) {
  if (cond) {
    process.stdout.write(`  PASS  ${label}\n`);
    passed += 1;
  } else {
    process.stdout.write(`  FAIL  ${label}\n`);
    failed += 1;
  }
}

const here = dirname(fileURLToPath(import.meta.url));
const frontendRoot = resolve(here, '..');

function source(rel: string): string {
  return readFileSync(resolve(frontendRoot, rel), 'utf8');
}

console.log('\nhermes fingerprint — desktop frontend');

// ── App.tsx — the #1 merge hotspot ──
{
  const s = source('App.tsx');

  // Hotbar digit-key handler (restored b6de38ac after 9-day gap)
  ok(s.includes('case "1"') && s.includes('case "7"'),
    'App.tsx: hotbar digit-key handler (cases 1-7)');
  ok(/case\s+"1":[\s\S]*?openPalette/.test(s),
    'App.tsx: key 1 → openPalette');
  ok(s.includes('setRightDockMode((cur) => cur === "context" ? "none" : "context")'),
    'App.tsx: key 5 dock toggle with "none" fallback');
  // RightDockMode type was moved to store/layout.ts by upstream refactoring
  ok(source('store/layout.ts').includes('"none"') && source('store/layout.ts').includes('type RightDockMode'),
    'store/layout.ts: RightDockMode includes "none"');
  // Hermes integration: the HermesSettings component is rendered via
  // SettingsPanel, not imported directly in App.tsx. The import lives in
  // SettingsPanel.tsx (verified separately below).
}

// ── SettingsPanel.tsx — Hermes tab injection ──
{
  const s = source('components/SettingsPanel.tsx');

  ok(s.includes('HermesLiveSection'),
    'SettingsPanel.tsx: HermesLiveSection component');
  ok(s.includes('useHermesLiveData'),
    'SettingsPanel.tsx: useHermesLiveData hook imported');
  ok(s.includes('HermesSettings') && s.includes('from "./hermes/HermesSettings"'),
    'SettingsPanel.tsx: HermesSettings imported from hermes/');
}

// ── bridge.ts — Hermes Wails bindings ──
{
  const s = source('lib/bridge.ts');

  ok(s.includes('Hermes bindings (restored after upstream merge)'),
    'bridge.ts: Hermes bindings block header');
  ok(s.includes('SetDesktopHotbar'),
    'bridge.ts: SetDesktopHotbar binding');
  ok(s.includes('SetProfiles'),
    'bridge.ts: SetProfiles binding');
  ok(s.includes('LearnedPatterns'),
    'bridge.ts: LearnedPatterns binding');
  ok(s.includes('CompressStats'),
    'bridge.ts: CompressStats binding');
  ok(s.includes('ScheduleDashboard'),
    'bridge.ts: ScheduleDashboard binding');
  ok(s.includes('CostSummary'),
    'bridge.ts: CostSummary binding');
  ok(s.includes('SyncLobeHubMarketplace'),
    'bridge.ts: SyncLobeHubMarketplace binding');
  ok(s.includes('hotbar:') && s.includes('profiles:') && s.includes('activeProfile:'),
    'bridge.ts: hotbar/profiles/activeProfile in mock defaults');
}

// ── types.ts — Hermes type definitions ──
{
  const s = source('lib/types.ts');

  ok(s.includes('HotbarView'),
    'types.ts: HotbarView interface');
  ok(s.includes('ProfileView'),
    'types.ts: ProfileView interface');
  ok(s.includes('hotbar: HotbarView'),
    'types.ts: hotbar field on SettingsView');
  ok(s.includes('activeProfile:'),
    'types.ts: activeProfile field on SettingsView');
}

// ── theme.ts — Hermes theme style ──
{
  const s = source('lib/theme.ts');

  ok(s.includes("'hermes'"),
    "theme.ts: 'hermes' in theme style definitions");
}

// ── StatusBar.tsx — Hermes status bar group ──
{
  const s = source('components/StatusBar.tsx');

  ok(s.includes('statusbar__group--hermes'),
    'StatusBar.tsx: statusbar__group--hermes class');
}

// ── index.html — window title ──
{
  const s = readFileSync(resolve(frontendRoot, '..', 'index.html'), 'utf8');

  ok(s.includes('Reasonix-Hermes'),
    'index.html: <title>Reasonix-Hermes</title>');
}

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
